File: sort/Quick.ex
defmodule Sort do
  def quick(arr, left, right) do
    {arr, index} = partition(arr, left, right)

    if (left < index - 1), do: arr = quick(arr, left, index - 1)
    if (index < right), do: arr = quick(arr, index, right)

    arr
  end

  def partition(arr, left, right) do
    i = left
    j = right
    pivot = Enum.at(arr, div(left + right, 2))

    compare(arr, i, j, pivot)
  end

  def compare(arr, i, j, pivot) do
    cond do
      Enum.at(arr, i) < pivot -> compare(arr, i+1, j, pivot)
      Enum.at(arr, j) > pivot -> compare(arr, i, j-1, pivot)
      true ->
        if (i <= j) do
          arr = swap(arr, i, j)
          compare(arr, i+1, j-1, pivot)
        else
          {arr, i}
        end
    end
  end

  def swap(arr, l, r) do
    {x, y} = {Enum.at(arr, l), Enum.at(arr, r)}

    arr = List.replace_at(arr, l, y)
    arr = List.replace_at(arr, r, x)
  end
end

array = Enum.shuffle(1..50) # [2, 3, 1, 5, 4]

array_sorted = Sort.quick(array, 0, length(array))

IO.inspect array_sorted

File: sort/Quick.cs
using System;
using System.Collections.Generic;
using System.IO;

class QuickSort 
{
  static void Main(String[] args) 
  {
    int[] unsorted = new int[]{2,3,1,5,4};
    int[] sorted = Quick(unsorted);
    
    foreach(var item in sorted)
    {
      Console.WriteLine(item);
    }
  }
  
  // Método utilizado apenas para envelopar a função recursiva (wrapper-function)
  static int[] Quick(int[] array)
  {
    return Quick(array, 0, array.Length - 1);
  }
  
  static int[] Quick(int[] array, int left, int right)
  {
    int index = Partition(array, left, right);
    
    if(left < index - 1) Quick(array, left, index - 1);
    if(index < right) Quick(array, index, right);
    
    return array;
  }
  
  static int Partition(int[] array, int left, int right)
  {
    int i = left,
        j = right,
        pivot = array[(left + right) / 2];
            
    while(i <= j)
    {
      while(array[i] < pivot) i++;
      while(array[j] > pivot) j--;
                    
      if(i <= j)
      {
        Swap(array, i, j);
        i++;
        j--;
      }
    }
            
    return i;
  }
  
  static void Swap(int[] array, int a, int b)
  {
    int aux = array[a];
    array[a] = array[b];
    array[b] = aux;
  }
}
